File: src/scoring/confidence.js
export function computeConfidence({ review, testResult, proposal, plan }) {
  let confidence = 0;
  const evidence = [];

  if (proposal?.diff?.trim() || proposal?.workingDiff?.trim()) {
    confidence += 0.2;
    evidence.push("diff generated");
  }

  if (plan?.context?.files?.length) {
    confidence += 0.1;
    evidence.push("graph-ranked context retrieved");
  }

  if (testResult?.results?.length) {
    const passed = testResult.results.filter((result) => result.ok).length;
    const ratio = passed / testResult.results.length;
    confidence += 0.35 * ratio;
    evidence.push(`${passed}/${testResult.results.length} validation commands passed`);
  }

  if (review?.score != null) {
    confidence += Math.max(0, Math.min(1, Number(review.score))) * 0.35;
    evidence.push(`review score ${review.score}`);
  }

  const blockers = review?.findings?.filter((finding) => finding.severity === "blocker").length ?? 0;
  if (blockers) {
    confidence = Math.min(confidence, 0.49);
    evidence.push(`${blockers} blocker finding(s)`);
  }

  return {
    confidence: Number(confidence.toFixed(3)),
    evidence,
    approved: confidence >= 0.8 && !blockers
  };
}

File: go_issue_agent/scoring/confidence.py
from __future__ import annotations

def compute_confidence(review: dict | None, test_result: dict | None, proposal: dict | None, plan: dict | None) -> dict:
    confidence = 0.0
    evidence = []

    if (proposal or {}).get("diff", "").strip() or (proposal or {}).get("workingDiff", "").strip():
        confidence += 0.2
        evidence.append("diff generated")

    if ((plan or {}).get("context") or {}).get("files"):
        confidence += 0.1
        evidence.append("graph-ranked context retrieved")

    results = (test_result or {}).get("results") or []
    if results:
        passed = len([result for result in results if result.get("ok")])
        ratio = passed / len(results)
        confidence += 0.35 * ratio
        evidence.append(f"{passed}/{len(results)} validation commands passed")

    if review and review.get("score") is not None:
        score = max(0.0, min(1.0, float(review["score"])))
        confidence += score * 0.35
        evidence.append(f"review score {review['score']}")

    blockers = len([finding for finding in (review or {}).get("findings", []) if finding.get("severity") == "blocker"])
    if blockers:
        confidence = min(confidence, 0.49)
        evidence.append(f"{blockers} blocker finding(s)")

    confidence = round(confidence, 3)
    return {"confidence": confidence, "evidence": evidence, "approved": confidence >= 0.8 and blockers == 0}
